File: ts-plugins/src/api/plugin-registry.ts
import type { FlowDefinition, FlowEngine, Builder } from '@unlaxer/tramli';
import type {
  FlowPlugin, AnalysisPlugin, StorePlugin, EnginePlugin,
  RuntimeAdapterPlugin, GenerationPlugin
} from './types.js';
import { PluginReport } from './types.js';

/**
 * Central registry for tramli plugins.
 * Manages plugin lifecycle: analyze → wrap store → install engine → bind adapters.
 */
export class PluginRegistry<S extends string> {
  private plugins: FlowPlugin[] = [];

  register(plugin: FlowPlugin): this {
    this.plugins.push(plugin);
    return this;
  }

  /** Run all analysis plugins against a FlowDefinition. */
  analyzeAll(definition: FlowDefinition<S>): PluginReport {
    const report = new PluginReport();
    for (const p of this.plugins) {
      if (p.kind() === 'ANALYSIS') {
        (p as AnalysisPlugin<S>).analyze(definition, report);
      }
    }
    return report;
  }

  /** Build a FlowDefinition and run all analysis plugins. Throws if any ERROR findings. */
  buildAndAnalyze(builder: Builder<S>): FlowDefinition<S> {
    const def = builder.build();
    const report = this.analyzeAll(def);
    const errors = report.findings().filter(f => f.severity === 'ERROR');
    if (errors.length > 0) {
      throw new Error(`Analysis errors:\n${errors.map(e => `  [${e.pluginId}] ${e.message}`).join('\n')}`);
    }
    return def;
  }

  /** Apply all store plugins (wrapping in registration order). */
  applyStorePlugins(store: any): any {
    let wrapped = store;
    for (const p of this.plugins) {
      if (p.kind() === 'STORE') {
        wrapped = (p as StorePlugin).wrapStore(wrapped);
      }
    }
    return wrapped;
  }

  /** Install all engine plugins. */
  installEnginePlugins(engine: FlowEngine): void {
    for (const p of this.plugins) {
      if (p.kind() === 'ENGINE') {
        (p as EnginePlugin).install(engine);
      }
    }
  }

  /** Bind all runtime adapter plugins. Returns map of plugin id → bound adapter. */
  bindRuntimeAdapters(engine: FlowEngine): Map<string, unknown> {
    const adapters = new Map<string, unknown>();
    for (const p of this.plugins) {
      if (p.kind() === 'RUNTIME_ADAPTER') {
        const adapter = (p as RuntimeAdapterPlugin<unknown>).bind(engine);
        adapters.set(p.descriptor().id, adapter);
      }
    }
    return adapters;
  }
}

File: java-plugins/src/main/java/org/unlaxer/tramli/plugins/api/PluginRegistry.java
package org.unlaxer.tramli.plugins.api;

import org.unlaxer.tramli.FlowDefinition;
import org.unlaxer.tramli.FlowEngine;
import org.unlaxer.tramli.FlowState;
import org.unlaxer.tramli.FlowStore;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class PluginRegistry {
    private final List<FlowPlugin> plugins = new ArrayList<>();

    public PluginRegistry register(FlowPlugin plugin) {
        plugins.add(plugin);
        return this;
    }

    public List<FlowPlugin> plugins() {
        return Collections.unmodifiableList(plugins);
    }

    public <S extends Enum<S> & FlowState> PluginReport analyzeAll(FlowDefinition<S> definition) {
        PluginReport report = new PluginReport();
        for (FlowPlugin plugin : plugins) {
            if (plugin instanceof AnalysisPlugin<?> raw) {
                @SuppressWarnings("unchecked")
                AnalysisPlugin<S> typed = (AnalysisPlugin<S>) raw;
                typed.analyze(definition, report);
            }
        }
        return report;
    }

    /** Build a FlowDefinition and run all analysis plugins. Throws if any ERROR findings. */
    public <S extends Enum<S> & FlowState> FlowDefinition<S> buildAndAnalyze(FlowDefinition.Builder<S> builder) {
        FlowDefinition<S> def = builder.build();
        PluginReport report = analyzeAll(def);
        var errors = report.findings().stream().filter(f -> "ERROR".equals(f.severity())).toList();
        if (!errors.isEmpty()) {
            var msg = errors.stream().map(e -> "  [" + e.pluginId() + "] " + e.message()).reduce("", (a, b) -> a + "\n" + b);
            throw new IllegalStateException("Analysis errors:" + msg);
        }
        return def;
    }

    /** Run all analysis plugins and throw if any ERROR findings. For already-built definitions. */
    public <S extends Enum<S> & FlowState> void analyzeAndValidate(FlowDefinition<S> definition) {
        PluginReport report = analyzeAll(definition);
        var errors = report.findings().stream().filter(f -> "ERROR".equals(f.severity())).toList();
        if (!errors.isEmpty()) {
            var msg = errors.stream().map(e -> "  [" + e.pluginId() + "] " + e.message()).reduce("", (a, b) -> a + "\n" + b);
            throw new IllegalStateException("Analysis errors:" + msg);
        }
    }

    public FlowStore applyStorePlugins(FlowStore baseStore) {
        FlowStore current = baseStore;
        for (FlowPlugin plugin : plugins) {
            if (plugin instanceof StorePlugin storePlugin) {
                current = storePlugin.wrapStore(current);
            }
        }
        return current;
    }

    public void installEnginePlugins(FlowEngine engine) {
        for (FlowPlugin plugin : plugins) {
            if (plugin instanceof EnginePlugin enginePlugin) {
                enginePlugin.install(engine);
            }
        }
    }

    public <R> List<R> bindRuntimeAdapters(FlowEngine engine, Class<R> adapterType) {
        List<R> adapters = new ArrayList<>();
        for (FlowPlugin plugin : plugins) {
            if (plugin instanceof RuntimeAdapterPlugin<?> rawAdapter) {
                Object bound = rawAdapter.bind(engine);
                if (adapterType.isInstance(bound)) {
                    adapters.add(adapterType.cast(bound));
                }
            }
        }
        return adapters;
    }
}
